Add Integer-valued Hashes

emilypi opened the pull request:
This addresses [#172](https://github.com/kadena-io/pact/issues/172)

emilypi commented on src/Pact/Types/Hash.hs:
```diff
@@ -33,3 +46,53 @@ verifyHash h b = if hash b == h
+  -> Text -- ^ The string to convert to integral base-a
+  -> Either Text Integer
+readBasedIntegerText base rChr txt
+  | base <= 1 = Left ("readStringAtBase: applied to unsupported base - " `append` asString base)
```
Yeah, I like it better as well. I'll change back.

emilypi commented on src/Pact/Types/Word512.hs:
```diff
@@ -0,0 +1,27 @@
+mkUnpackedDoubleWord "Word512" ''Word256 "Int512" ''Int256 ''Word256
+  [''Typeable, ''Data, ''Generic]
+
+
```
@slpopejoy we talked about needing a `Word512` type potentially - this would do it. If we need it, great! if not i can just remove this.

emilypi commented on src/Pact/Types/Hash.hs:
```diff
@@ -33,3 +46,53 @@ verifyHash h b = if hash b == h
+-- | Reads 'String' as a non-negative 'Integral' number using the base
+-- specified by the first argument, and character representation
+-- specified by the second argument
+readBasedIntegerText
```
There will most definitely be tests 😄

emilypi commented:
@mightybyte Gitlab is struggling a bit with Nix

fosskers commented:
@emilypi that's due to `megaparsec` in the short-term, but the non-deterministic hash test failures in the long-term.

emilypi commented:
@fosskers fun!!

joelburget commented:
cc @bts we should add analysis support once this is finalized

slpopejoy commented:
I'm beginning to wonder if this makes any sense. 

While we're used to hashes being `Int`s in Haskell, this is because they are not crypto-hashes like sha256 or blake2b 512; they fit comfortably into `Int`.

Meanwhile Eth has popularized "hash as number" because they decided for no good reason to make `Word256` their base type of everything. The one and only advantage you get is yay, public keys and hashes are primitives. For everything else it's a disaster.

The actual use of crypto hashes is as bytestrings, NOT as gargantuan integrals that won't fit in any math coprocessor. 

Therefore, I propose we switch gears and offer the ability to convert a hex string to an array of reasonably-sized pact `integer`s. This way we're not locked into the current 512-bit size (which is increasingly seeming absurd for application use, 256 is more than large enough). The most straightforward is to do bytes, thus
`(hex-to-array "ffffff")` => `[256,256,256]`, and so forth. You still have all the randomness but now we're not doing insane things like storing gargantuan bignums in the database.

We can obviously go the other way too, `(array-to-hex [256,256])` -> `"ffff"`

emilypi commented:
@slpopejoy That makes sense and I can definitely switch course and get rid of the code in this PR, unless you still think it has some value maybe eventually

mightybyte commented:
@slpopejoy That makes sense to me.  I was originally thinking of justing using `Integer` for convenience, but you make a good point about DB storage.  I think I would lean toward `[Int64]` or something like that.  As long as it's a usable integral I don't think it matters too much.  `[Int8]` seems a little on the small side though.

@emilypi I don't know that you need to totally ditch this PR.  It seems like just a matter of changing the output type.

emilypi commented:
@mightybyte i've added the rest of what we need here, if you could take a look 😄 

Since i've added `digit` as a dependency, it might be in our interest using it for lensy parsing as well, so i've added `Traversal'`s for Text-like numeric representations of things, should we need it. It's really slick to use.
